presto-kudu/src/test/java/io/prestosql/plugin/kudu/TestKuduIntegrationGroupedExecution.java

```diff
@@ -0,0 +1,207 @@
+                "key4 BIGINT WITH (primary_key=true)," +
+                "attr1 BIGINT" +
+                ") WITH (" +
+                " partition_by_hash_columns = ARRAY['key1', 'key2']," +
+                " partition_by_hash_buckets = 2," +
+                " partition_by_second_hash_columns = ARRAY['key3']," +
+                " partition_by_second_hash_buckets = 2" +
+                ")");
+
+        assertUpdate("INSERT INTO test_grouped_execution_mtlvl VALUES (0, 0, 0, 0, 0), (0, 0, 0, 1, 1), (1, 1, 1, 0, 0), (1, 1, 1, 1, 1)", 4);
+        assertQuery(groupedExecutionSessionProperties, "SELECT key1, key2, key3, COUNT(1) FROM test_grouped_execution_mtlvl GROUP BY key1, key2, key3", "SELECT * FROM (VALUES (0, 0, 0, 2), (1, 1, 1, 2))", assertRemoteExchangesCount(1));
+
+        assertUpdate("DROP TABLE test_grouped_execution_mtlvl");
+    }
+
+    @Test
+    public void testGroupedExecutionMultiLevelCombinedPartitioning()
+    {
+        Session groupedExecutionSessionProperties = Session.builder(getSession())
+                .setSystemProperty(COLOCATED_JOIN, "true")
+                .setSystemProperty(GROUPED_EXECUTION, "true")
+                .setSystemProperty(CONCURRENT_LIFESPANS_PER_NODE, "1")
+                .setSystemProperty(DYNAMIC_SCHEDULE_FOR_GROUPED_EXECUTION, "false")
+                .setSystemProperty(ENABLE_DYNAMIC_FILTERING, "false")
+                .setSystemProperty(KUDU_GROUPED_EXECUTION, "true")
+                .build();
+
+        assertUpdate("CREATE TABLE test_grouped_execution_hash_range (" +
+                "key1 BIGINT WITH (primary_key=true)," +
+                "key2 BIGINT WITH (primary_key=true)," +
+                "key3 BIGINT WITH (primary_key=true)," +
+                "key4 BIGINT WITH (primary_key=true)," +
+                "attr1 BIGINT" +
+                ") WITH (" +
+                "  partition_by_hash_columns = ARRAY['key1']," +
+                "  partition_by_hash_buckets = 2," +
+                "  partition_by_second_hash_columns = ARRAY['key2']," +
+                "  partition_by_second_hash_buckets = 3," +
+                "  partition_by_range_columns = ARRAY['key3']," +
+                "  range_partitions = '[{\"lower\": null, \"upper\": \"4\"}, {\"lower\": \"4\", \"upper\": \"9\"}, {\"lower\": \"9\", \"upper\": null}]'" +
+                ")");
+
+        assertUpdate("INSERT INTO test_grouped_execution_hash_range VALUES (0, 0, 0, 0, 0), (0, 0, 9, 0, 9), (0, 0, 9, 1, 0), (1, 1, 0, 0, 1), (1, 1, 9, 0, 2)", 5);
+        assertQuery(groupedExecutionSessionProperties, "SELECT key1, key2, key3, COUNT(1) FROM test_grouped_execution_hash_range GROUP BY key1, key2, key3", "SELECT * FROM (VALUES (0, 0, 0, 1), (0, 0, 9, 2), (1, 1, 0, 1), (1, 1, 9, 1))", assertRemoteExchangesCount(1));
+
+        assertUpdate("DROP TABLE test_grouped_execution_hash_range");
+    }
+
+    private Consumer<Plan> assertRemoteExchangesCount(int expectedRemoteExchangesCount)
```

Comment:
What this assertion proves? Why you expect certain number of exchange stages?

Reply:
If grouped_execution worked, there will be only one remoteExchange in group by and join operation, or there be bigger than 1.  :)